### ISSUE TITLE
Vulnerability Detector - Improve the translation mechanism to allow more complex version fields 

### ISSUE THREAD

**Issue by pereyra-m** [wazuh/wazuh#23656]
## Description
The translation functionality should be expanded to cover a wider spectrum of solutions.

Using the translation schema as a baseline:
```json
{
  "target": [
    "Platform 1",
    "Platform n"
  ],
  "source": {
    "vendor": "vendor name",
    "product": "package name",
    "version": "package version"
  },
  "translation": [
    {
      "vendor": "translated vendor name",
      "product": "translated package name",
      "version": "translated package version"
    }
  ],
  "action": [
    "replace_vendor",
    "set_version_if_product_matches",
    "replace_product"
  ]
}
```

## Definitions
- Add a new "extended" translations
  - This is going to be a [new resource](https://github.com/wazuh/wazuh/blob/623091670c4b2c201e93b875a7dfd63440c75169/src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/eventDecoder.hpp#L54-L58), under the name "extended-TID-xxx" 
- Maintain older translation functionality
- Move the business logic to a separate class (out from the databaseFeedManager and package scanner)

>[!NOTE]
> Take functionality extension into account, for the addition of new fields in both "translation" and "source" objects. eg. "format should be deb".

### Source
- Could contain 0 to N capture groups on each field
- The version field is optional
  - If the translation doesn't need to check the version value, it won't be provided

### Translation
- All fields are optional
  - If not provided, the value from syscollector is used instead
- If provided, should be constructed with placeholders, hardcoded strings or both
  - Eg: "$(VERSION_0)$(PRODUCT_2)$(PRODUCT_3)"
  - Eg: "NEW VALUE"
  - Eg: "1.$(VENDOR_1)_100"

### Other requirements
- Don't use flatbuffers in the extended implementation
  - It proved to be difficult to use and increased the risk of errors
  - Given the amount of translations available it doesn't necessarily improve memory usage by a significant factor (while degrading performance related to the parsing)
  - It forced us to use a hard schema, which made the implementation of this functionality much harder
- Simplify the translations:
  - Remove unused fields or re-implement in a generic way
    - action array
    - update
    - target_hw
    - msu_name
    - sw_edition

## Tasks
Create multiple PRs and point them to an intermediary branch

### Wazuh
- [ ] Move the business logic to a new "translation" class
  - [ ] Add an interface
  - [ ] Add the "legacy" translation functionality
  - [ ] Changes tested
- [ ] Add the new functionality
  - [ ] Add the extended translation functionality
  - [ ] Changes tested

### ID
- [ ] Translation adaptation
  - [ ] Remove unused fields
  - [ ] Add extended translations
  - [ ] Test integration

## DoD
- [ ] Translation documentation updated
- [ ] Translations updated
- [ ] Feature implemented
- [ ] Testing updated 
  - [ ] UTs
  - [ ] Component

**Comment by sebasfalcone** [comment]
I will discuss this with @Dwordcito in order to determine if its viable or not

**Comment by pereyra-m** [comment]
## Update

I begin with the implementation, I'll take the branch of #23518 as a starting point for this issue.

**Comment by pereyra-m** [comment]
## Update 

The issue is on hold due to more priority issues related to 4.8.0 release testing.

**Comment by sebasfalcone** [comment]
## Issue blocked
Awaiting merge of:
- https://github.com/wazuh/wazuh/issues/23518

**Comment by pereyra-m** [comment]
## Update 

Now the issue is no longer blocked, I'll start to work on it again.

**Comment by pereyra-m** [comment]
## Update

It was found that simply giving the manager the possibility of construct custom versions applying a regex to the name won't solve all the different cases that we could find.

A new approach will be tested, and the conclusions will be shared below.

**Comment by pereyra-m** [comment]
## Update

I've working on this issue, here I share some key points:
- Simply adding the capability of creating custom version strings using a regex extracted from the name won't solve all possible situations. A more generic approach is required.
- We cannot still differentiate among CPEs with the same `vendor` and `product`, but different `update`

### Example with Java

This Windows program requires applying a regex to the name and extracting two different groups: `version` and `update`.

![2024-06-18_12-33](https://github.com/wazuh/wazuh/assets/65046601/91123e6a-523a-4d71-8245-a283f52f560f)

If we don't consider the `update` field, a CVE like the following will end up having false positives because all `v1.x.0` will be reported as vulnerable

https://nvd.nist.gov/vuln/detail/CVE-2013-1480

### Example with Tomcat

Here we need to apply a regex to the version field, not the name, to obtain both the right version (to include the patch) and the milestone (and to convert its format to add the word "milestone")

![2024-06-18_12-35](https://github.com/wazuh/wazuh/assets/65046601/63de4472-e0e8-4acc-a22b-3e93eeb5a2b1)

For example, in the CVE https://nvd.nist.gov/vuln/detail/CVE-2016-8735


## Proposal

We could add a generic replacement like this one

<details><summary>Details</summary>
<p>

```c++
constexpr auto PRODUCT_PLACEHOLDER {"$PRODUCT"};
constexpr auto VENDOR_PLACEHOLDER {"$VENDOR"};
constexpr auto VERSION_PLACEHOLDER {"$VERSION"};

...
// Append the matching translation to the result vector
                for (const auto& translatedPackage : cacheData.translation)
                {
                    TranslatedData translatedResult;
                    translatedResult.translatedProduct = translatedPackage.translatedProduct;
                    translatedResult.translatedVendor = translatedPackage.translatedVendor;
                    translatedResult.translatedVersion = translatedPackage.translatedVersion;
                    translatedResult.translatedUpdate = translatedPackage.translatedUpdate;

                    auto placeHolderReplacement =
                        [&translatedResult](auto extractedGroups, auto placeholder)
                    {
                        if (extractedGroups.size() > 1)
                        {
                            for (size_t i = 1; i < extractedGroups.size(); i++)
                            {
                                Utils::replaceAll(translatedResult.translatedProduct,
                                                  extractedGroups + std::to_string(i),
                                                  extractedGroups.str(i));
                                Utils::replaceAll(translatedResult.translatedVendor,
                                                  extractedGroups + std::to_string(i),
                                                  extractedGroups.str(i));
                                Utils::replaceAll(translatedResult.translatedVersion,
                                                  extractedGroups + std::to_string(i),
                                                  extractedGroups.str(i));
                                Utils::replaceAll(translatedResult.translatedUpdate,
                                                  extractedGroups + std::to_string(i),
                                                  extractedGroups.str(i));
                            }
                        }
                    }

                    placeHolderReplacement(productExtractedGroups, PRODUCT_PLACEHOLDER);
                    placeHolderReplacement(vendorExtractedGroups, VENDOR_PLACEHOLDER);
                    placeHolderReplacement(versionExtractedGroups, VERSION_PLACEHOLDER);

                    translationResult.push_back(std::move(translatedResult));
                }

```

</p>
</details> 

We apply a regex to the product, vendor, and version, and then replace all groups found in all fields. This way, we'd be able to create generic translations for all cases.

But the `update` field is still an issue, we can either:
- Change the way we generate the candidates, to add to the NVD ones the `update` field from the CPE
- Update the content, and append to the name of the product, the `update` field. Then, we consider some rules for the translations, like "PRODUCT-UPDATE", "PRODUCT_UPDATE", etc.

**Comment by pereyra-m** [comment]
## Update

A draft of a possible implementation can be found at
https://github.com/wazuh/wazuh/tree/enhancement/23656-complex-version-fields

And the translations format is presented at
https://github.com/wazuh/intelligence-data/tree/enhancement/23656-complex-version-fields

**Comment by sebasfalcone** [comment]
Great work @pereyra-m! The problem was more complex than expected so we will continue with a baseline investigation first

## Issue blocked
We need to research the implications of the update field on the content and its correlation with syscollector information

**Comment by pereyra-m** [comment]
## Update

Now that issue is no longer blocked, I continue with the issue and remove from the implementation all the extra changes as decided in the meetings.

**Comment by pereyra-m** [comment]
## Update

I removed all extra changes and only left those related to the translation, I'm adding UT.
Missing steps: check and correct all translations, add QA integration tests, and solve how to merge this breaking change.

**Comment by pereyra-m** [comment]
## Update

I verify the QA tests with the modified translations locally to make sure they work with the new changes.

**Comment by pereyra-m** [comment]
## Update

I resume the issue. I'll create the required sanitization issues for the affected packages

**Comment by pereyra-m** [comment]
## Update

On-hold until the spike is completed.

**Comment by sebasfalcone** [comment]
## Issue blocked
Spike needed to continue development:
- https://github.com/wazuh/intelligence-platform/issues/1663

**Comment by sebasfalcone** [comment]
## ETA updated
ETA updated to match spike ETA

**Comment by sebasfalcone** [comment]
## Update
We synced the status of the related issues @Damian-Mangold. We will revisit this issue once the first spike starts

**Comment by sebasfalcone** [comment]
## Update
We synced the status of the related issues @Damian-Mangold. This issue is ready to work again

## Issue blocked
Awaiting for issue new definitions

**Comment by sebasfalcone** [comment]
## ETA update
We need to re think this issue from scratch

@Dwordcito I will move it to triage and re-write it again

**Comment by sebasfalcone** [comment]
## Definition update
With @pereyra-m we updated the definitions of this issue

Triage should be performed again

**Comment by pereyra-m** [comment]
## Update

I begin with the initial test to confirm the effects of the new translations over the 4.x code.

**Comment by sebasfalcone** [comment]
## Continuation
@pereyra-m Thanks for the testing!

We are going to have another consumer in 5.0.0, which will allow us to have distinguished translations between 4.x and 5.x and above

**Comment by sebasfalcone** [comment]
## Issue blocked
Meeting pending to define target branch for the translations and maintenance plan

**Comment by sebasfalcone** [comment]
## Issue blocked
Issue definitions update pending

**Comment by pereyra-m** [comment]
@Dwordcito 

To unblock this issue, we should define if v4.13.0 will support this type of translation or if it'll be available for 5.0

**Comment by pereyra-m** [comment]
Hi @Damian-Mangold 
I understand that the scanner doesn't have access to that field. And the proposed translation mechanism doesn't cover that case either.

**Comment by Damian-Mangold** [comment]
List of blocked translations awaiting this functionality:
 - https://github.com/wazuh/intelligence-data/issues/927
 - https://github.com/wazuh/intelligence-data/issues/2165
 - https://github.com/wazuh/intelligence-data/issues/2166
 - https://github.com/wazuh/intelligence-data/issues/2257
 - https://github.com/wazuh/intelligence-data/issues/2436

**Comment by Dwordcito** [comment]
if we have 4.15, we will introduce these changes here. If not, we will introduce this in 5.0

**Comment by jepalfer** [comment]
# 18/02/2026

Today I've been investigating what this issue is attempting to solve, reading all the commentaries leaved above, and investigating the translation workflow.

**Comment by MiguelazoDS** [comment]
# Analysis 

I've been going through the analysis and the root cause of this implementation. I need to take a deeper look and understand how many packages may be affected by these changes. 

Useful information can be found here too: https://github.com/wazuh/external-devel-requests/issues/1327#issuecomment-1863583396 

We must consider at least this 4 cases 

```console
- cpe:2.3:a:sun:jdk:*:*:*:*:*:*:*:*
- cpe:2.3:a:sun:jre:*:*:*:*:*:*:*:* 
- cpe:2.3:a:oracle:jdk:*:*:*:*:*:*:*:*
- cpe:2.3:a:oracle:jre:*:*:*:*:*:*:*:*
```

And if possible verify what's this: 

```console
- cpe:2.3:a:hp:jdk:*:*:*:*:*:*:*:*
- cpe:2.3:a:hp:jre:*:*:*:*:*:*:*:*
```

## Different formats

Depending on the version, we can find multiple different formats representing the version for jdk, and jre. 

### JDK and JRE Oracle and Sun 

https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&keyword=cpe%3A2.3%3Aa%3Asun%3Ajdk%3A*%3A*%3A*%3A*%3A*%3A*%3A*%3A*
https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&keyword=cpe%3A2.3%3Aa%3Asun%3Ajre%3A*%3A*%3A*%3A*%3A*%3A*%3A*%3A*
https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajdk%3A*%3A*%3A*%3A*%3A*%3A*%3A*%3A*
https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A*%3A*%3A*%3A*%3A*%3A*%3A*%3A*

We can found the following different but not limited to:

```console
cpe:2.3:a:sun:jdk:1.1.6:update7 
cpe:2.3:a:oracle:jdk:1.4.2_38:*:*:*:*:*:*:* 
cpe:2.3:a:oracle:jre:1.5.0:update4:*:*:*:*:*:* 
cpe:2.3:a:sun:jdk:1.6.0:update_6:*:*:*:*:*:* 
cpe:2.3:a:oracle:jdk:1.6.0:update115:*:*:*:*:*:*
```

We can see:

- Mixing `update` keyword with and without `_` before number
- Use of `update` keyword before version 4 and after version 4
- Sun and Oracle vendors exist simultaneously for certain version (I assume it was during the transition)
- Single number version with `_`   

### JDK and JRE HP 

- We only have:

```console
[cpe:2.3:a:hp:jdk:7.0.08:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/924727?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Ahp%3Ajdk%3A*%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL)
```

- In both cases the CVEs are:

```console
CVE-2014-0418
CVE-2014-0382
CVE-2013-5906
CVE-2013-5904
CVE-2013-5895
CVE-2013-5870
```

### Tomcat 

https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL&startIndex=0

For this package we can see 

```console
[cpe:2.3:a:apache:tomcat:3.0:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/14175?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:3.2.2:beta2:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/14181?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:4:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/122830?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:6.0.0:alpha:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/160500?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:7.0.0:rc1:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/615906?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:8.0.53-29.32.1:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/788315?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:9.0.0:milestone1:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/615922?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:9.0.1:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/351827?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:9.0.35-3.39.1:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/788316?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
[cpe:2.3:a:apache:tomcat:10.0.0:milestone1:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/detail/706876?namingFormat=2.3&orderBy=CPEURI&keyword=tomcat&status=FINAL)
```

For this case I don't see clearly if we need this change since it seems cpe:2.3:a:apache:tomcat:9.0.0:-:*:*:*:*:*:* covers all the milestone cases. 

https://nvd.nist.gov/vuln/detail/CVE-2025-61795#range-20488579

**Comment by MiguelazoDS** [comment]
# Update 04/08 

These are the packages I could install related to Java, multiple JRE and JDK versions 

```console
Java 9.0.1 (64-bit)
Java 7 Update 71 (64-bit)
Java 8 Update 111 (64-bit)
Java(TM) 6 Update 27 (64-bit)
Java(TM) 6 Update 45 (64-bit)
Java(TM) SE Development Kit 21.0.9 (64-bit)
Java(TM) SE Development Kit 9 (64-bit)
Java(TM) SE Development Kit 6 Update 45 (64-bit)
Java SE Development Kit 8 Update 111 (64-bit)
Java 8 Update 481 (64-bit)
Java(TM) SE Development Kit 25.0.1 (64-bit)
Java(TM) SE Development Kit 17.0.12 (64-bit)
Java 9 (64-bit)
Java(TM) SE Development Kit 15 (64-bit)
J2SE Runtime Environment 5.0 Update 4
Java Auto Updater
Java 8 Update 481
```

[packages.json](https://github.com/user-attachments/files/26572403/packages.json)

I will implement some base testing cases to validate later the changes do not break the logic.

**Comment by MiguelazoDS** [comment]
# Update 04/09 

I installed these 17 Java packages in Windows Server 2019 to verify the format for multiple versions. 

[java_packages.json](https://github.com/user-attachments/files/26612247/java_packages.json)

Additionally these other 4 packages 

<details><summary>Expand</summary>

```json
{
  "architecture": "x86_64",
  "category": null,
  "description": null,
  "installed": "2023-11-28T21:08:00+00:00",
  "multiarch": null,
  "name": "Java(TM) SE Development Kit 17.0.8 (64-bit)",
  "path": "C:\\Program Files\\Java\\jdk-17\\",
  "priority": null,
  "size": 0,
  "source": null,
  "type": "win",
  "vendor": "Oracle Corporation",
  "version": "17.0.8.0"
}
```

```json
{
  "architecture": "x86_64",
  "category": null,
  "description": null,
  "installed": "2023-12-14T15:43:50+00:00",
  "multiarch": null,
  "name": "Java 8 Update 351 (64-bit)",
  "path": "C:\\Program Files\\Java\\jre1.8.0_351\\",
  "priority": null,
  "size": 0,
  "source": null,
  "type": "win",
  "vendor": "Oracle Corporation",
  "version": "8.0.3510.10"
}
```

```json
{
  "architecture": "x86_64",
  "category": null,
  "description": null,
  "installed": "2023-12-19T13:55:33+00:00",
  "multiarch": null,
  "name": "Java SE Development Kit 7 Update 45 (64-bit)",
  "path": "C:\\Program Files\\Java\\jdk1.7.0_45\\",
  "priority": null,
  "size": 0,
  "source": null,
  "type": "win",
  "vendor": "Oracle",
  "version": "1.7.0.450"
}
```

```json
{
  "architecture": "i686",
  "category": null,
  "description": null,
  "installed": "2024/05/14 14:12:38",
  "multiarch": null,
  "name": "Java 2 Runtime Environment Standard Edition v1.3.1",
  "path": null,
  "priority": null,
  "size": 0,
  "source": null,
  "type": "win",
  "vendor": null,
  "version": null
}
```

</details>

After removing duplicates and redundant entries, I created the following reference table. 

| name | version | vendor | CPE | CVE example |
|---|---|---|---|---|
| Java(TM) SE Development Kit 17.0.8 (64-bit) | 17.0.8.0 | Oracle Corporation | [cpe:2.3:a:oracle:jdk:17.0.8:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajdk%3A17.0.8%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2023-22081](https://nvd.nist.gov/vuln/detail/CVE-2023-22081) |
| Java 8 Update 351 (64-bit) | 8.0.3510.10 | Oracle Corporation | [cpe:2.3:a:oracle:jre:1.8.0:update351:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A1.8.0%3Aupdate351%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2023-21843](https://nvd.nist.gov/vuln/detail/CVE-2023-21843) |
| Java SE Development Kit 7 Update 45 (64-bit) | 1.7.0.450 | Oracle | [cpe:2.3:a:oracle:jdk:1.7.0:update45:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajdk%3A1.7.0%3Aupdate45%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2020-2801](https://nvd.nist.gov/vuln/detail/CVE-2020-2801) |
| Java 2 Runtime Environment Standard Edition v1.3.1 | | | [cpe:2.3:a:sun:jre:1.3.1:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Asun%3Ajre%3A1.3.1%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2012-1720](https://nvd.nist.gov/vuln/detail/CVE-2012-1720) |
| Java 9.0.1 (64-bit) | 9.0.1.0 | Oracle Corporation | [cpe:2.3:a:oracle:jre:9.0.1:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A9.0.1%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2018-2678](https://nvd.nist.gov/vuln/detail/CVE-2018-2678) |
| Java 7 Update 71 (64-bit) | 7.0.710 | Oracle | [cpe:2.3:a:oracle:jre:1.7.0:update71:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A1.7.0%3Aupdate71%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2015-0410](https://nvd.nist.gov/vuln/detail/CVE-2015-0410) |
| Java(TM) 6 Update 27 (64-bit) | 6.0.270 | Oracle | [cpe:2.3:a:oracle:jre:1.6.0:update27:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A1.6.0%3Aupdate27%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2013-5852](https://nvd.nist.gov/vuln/detail/CVE-2013-5852) |
| Java(TM) SE Development Kit 9 (64-bit) | 9.0.0.0 | Oracle Corporation | [cpe:2.3:a:oracle:jdk:9:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajdk%3A9%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2022-22965](https://nvd.nist.gov/vuln/detail/CVE-2022-22965) |
| Java(TM) SE Development Kit 6 Update 45 (64-bit) | 1.6.0.450 | Oracle | [cpe:2.3:a:oracle:jdk:1.6.0:update45:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajdk%3A1.6.0%3Aupdate45%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2013-5852](https://nvd.nist.gov/vuln/detail/CVE-2013-5852) |
| Java 9 (64-bit) | 9.0.0.0 | Oracle Corporation | [ cpe:2.3:a:oracle:jre:9:*:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Aoracle%3Ajre%3A9%3A*%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | |
| J2SE Runtime Environment 5.0 Update 4 | 1.5.0.40 | Sun Microsystems, Inc. | [cpe:2.3:a:sun:jre:1.5.0:update4:*:*:*:*:*:*](https://nvd.nist.gov/products/cpe/search/results?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Asun%3Ajre%3A1.5.0%3Aupdate4%3A*%3A*%3A*%3A*%3A*%3A*&status=FINAL) | [CVE-2013-5850](https://nvd.nist.gov/vuln/detail/CVE-2013-5850) |
| Java Auto Updater | 9.0.1.0 | Oracle Corporation | | |

> [!NOTE]
> From Java 9, the update and the leading `1.` is no longer used https://openjdk.org/jeps/223 as we can observe for the first and third elements in the table. 

Update field in CPE is not used in the correlation matching (i.e Java 8 Update 351 (64-bit))

```json
{"jre_CVE-2023-21843": {
  "candidates": [
    {
      "cveId": "CVE-2023-21843",
      "defaultStatus": "unaffected",
      "versions": [
        {
          "version": "1.8.0"
        },
        {
          "version": "11.0.17"
        },
        {
          "version": "17.0.5"
        },
        {
          "version": "19.0.1"
        }
      ],
      "vendor": "oracle"
    }
  ]
}
```

I don't find any issue in matching the vendor, it seems it is always is either "Oracle" or "Oracle Corporation" with less cases of "Sun" related to versions prior 2010. It was a period of time we can find either Oracle or Sun vendor during the transition for the same version. 

Currently from the 12 packages, only 6 match (although those matches are not all correct) 

```console
scanPackageTranslation():No translation found for 'Java 9 (64-bit)', using original name
scanPackageTranslation():Translation L2 cache hit for 'Java(TM) SE Development Kit 6 Update 45 (64-bit)' (vendor: 'Oracle') on 'windows', promoting to L1
scanPackageTranslation():Translation L2 cache hit for 'Java(TM) SE Development Kit 9 (64-bit)' (vendor: 'Oracle Corporation') on 'windows', promoting to L1
scanPackageTranslation():No translation found for 'Java(TM) 6 Update 27 (64-bit)', using original name
scanPackageTranslation():No translation found for 'Java 7 Update 71 (64-bit)', using original name
scanPackageTranslation():No translation found for 'Java 9.0.1 (64-bit)', using original name
scanPackageTranslation():Translation L2 cache hit for 'Java 2 Runtime Environment Standard Edition v1.3.1' (vendor: ' ') on 'windows', promoting to L1
scanPackageTranslation():Translation L2 cache hit for 'Java SE Development Kit 7 Update 45 (64-bit)' (vendor: 'Oracle') on 'windows', promoting to L1
scanPackageTranslation():No translation found for 'Java Auto Updater', using original name
scanPackageTranslation():No translation found for 'J2SE Runtime Environment 5.0 Update 4', using original name
scanPackageTranslation():Translation L2 cache hit for 'Java 8 Update 351 (64-bit)' (vendor: 'Oracle Corporation') on 'windows', promoting to L1
scanPackageTranslation():Translation L2 cache hit for 'Java(TM) SE Development Kit 17.0.8 (64-bit)' (vendor: 'Oracle Corporation') on 'windows', promoting to L1
```

The issue here is that this case 

https://nvd.nist.gov/vuln/detail/CVE-2013-5852

Does not affect update above 60, but according to the feed information everything with 1.6.0 is affected.

**Comment by MiguelazoDS** [comment]
# Update 04/13 

Validate the implementation on this PR and notify the content team to prevent false positives. The format could be 1.6.0.45, otherwise either the vulnerability won't be detected because it will always be greater than 1.6.0, or we will have multiple false positives reporting any 1.6.0 version without considering the update. 

This is another case that needs to be fixed (1.4.2_37)

```json
jre_CVE-2012-1720 ==> {
  "candidates": [
    {
      "cveId": "CVE-2012-1720",
      "defaultStatus": "unaffected",
      "versions": [
        {
          "version": "0",
          "lessThanOrEqual": "1.6.0",
          "versionType": "custom"
        },
        {
          "version": "0",
          "lessThanOrEqual": "1.7.0",
          "versionType": "custom"
        }
      ],
      "vendor": "oracle"
    },
    {
      "cveId": "CVE-2012-1720",
      "defaultStatus": "unaffected",
      "versions": [
        {
          "version": "0",
          "lessThanOrEqual": "1.4.2_37",
          "versionType": "custom"
        },
        {
          "version": "0",
          "lessThanOrEqual": "1.5.0",
          "versionType": "custom"
        }
      ],
      "vendor": "sun"
    }
  ]
}
```

<details><summary>Expand</summary>

<img width="556" height="380" alt="Image" src="https://github.com/user-attachments/assets/1c69ddf8-677c-4b11-980d-6e34585c7aa4" />
<img width="653" height="416" alt="Image" src="https://github.com/user-attachments/assets/00ebd373-23e3-4b57-aaa8-a62de3269f10" />
<img width="800" height="445" alt="Image" src="https://github.com/user-attachments/assets/5fb4a09f-006e-4846-8f66-56f678520261" />
<img width="790" height="447" alt="Image" src="https://github.com/user-attachments/assets/9c0cda50-b0ef-4ef2-8b91-f49bcc92c4ab" />

</details>

## Update 

Testing the new translations with older logic, the new format is not completely backward compatible. I minor change was done by using the "update" field for "version", since that field is not used in 4.x it should be fully compatible.

**Comment by MiguelazoDS** [comment]
# Update 

After the research, summarized here https://github.com/wazuh/wazuh/issues/36398#issuecomment-4692952712, I'll start implementing the changes and efficacy tests. The last ones, will be addressed using an offline feed but this should be moved to a more suitable place (waiting definitions). 

Addressing the "trailing zero issue" 

Windows package
https://github.com/wazuh/wazuh/issues/36093

for 

```console
visual_studio_tools_for_applications_2019
visual_studio_tools_for_applications_2019_sdk
visual_studio_tools_for_applications_2022
visual_studio_tools_for_applications_2022_sdk
```

... and probably many other packages 

NVD reports "16.0.35907.0", while Syscollector data provider do not include the trailing zero. 

Ubuntu package (PPA that uses NVD) 
https://github.com/wazuh/wazuh/issues/36731

Not only the dpkg compare version issue, but also the a user reported that Syscollector data provider is sending wrong information (151.0 instead of 151.0.1). 

I was not able to demonstrate that, but what it is clear that is very common to have trailing zeros for Windows packages. 

<details><summary> Expand </summary> 

```json
{
      "architecture": "x86_64",
      "description": " ",
      "format": "win",
      "groups": " ",
      "install_time": "2025/08/26 14:13:46",
      "location": " ",
      "name": "Mozilla Maintenance Service",
      "priority": " ",
      "size": 0,
      "source": " ",
      "vendor": "Mozilla",
      "version": "142.0"
    },
{
      "architecture": "x86_64",
      "description": " ",
      "format": "win",
      "groups": " ",
      "install_time": "2025/04/28 02:46:55",
      "location": " ",
      "name": "Oracle VM VirtualBox Guest Additions 6.1.22",
      "priority": " ",
      "size": 0,
      "source": " ",
      "vendor": "Oracle Corporation",
      "version": "6.1.22.0"
    },
{
      "architecture": "x86_64",
      "description": " ",
      "format": "win",
      "groups": " ",
      "install_time": "2026/01/15 21:27:25",
      "location": " ",
      "name": "Microsoft OLE DB Driver for SQL Server",
      "priority": " ",
      "size": 0,
      "source": " ",
      "vendor": "Microsoft Corporation",
      "version": "18.7.4.0"
    },
{
      "architecture": "x86_64",
      "description": " ",
      "format": "win",
      "groups": " ",
      "install_time": "2021/05/11 01:00:00",
      "location": "c:\\Program Files\\Microsoft Silverlight\\",
      "name": "Microsoft Silverlight",
      "priority": " ",
      "size": 0,
      "source": " ",
      "vendor": "Microsoft Corporation",
      "version": "5.1.50918.0"
    }
``` 

</details> 

> [!NOTE]
>Both cases are possible, trailing zeros in syscollector packages and trailing zeros in NVD content.

**Comment by MiguelazoDS** [comment]
# Update 

Implementing integration tests for trailing zero issue in NVD feed, leads to errors in the local environment. 

<details><summary>Expand</summary> 

```
start():Schema validator initialized successfully from embedded resources.
readGlobalMapsNoLock():Global maps are not yet available in the local feed database, skipping reload.
TDatabaseFeedManager():No existing local vulnerability feed detected at startup. A full feed will be downloaded.
start():CVE feed not yet available — per-agent scans deferred until initial load completes.
start():Vulnerability scanner module started.
operator()():Initiating feed update process.
operator()():Error updating feed: Unable to parse payload: 1: 11: error: unknown field: action.
operator()():Initiating feed update process.
operator()():Error updating feed: Unable to parse payload: 1: 11: error: unknown field: action.
operator()():Initiating feed update process.
operator()():Error updating feed: Unable to parse payload: 1: 11: error: unknown field: action.
operator()():Initiating feed update process.
operator()():Error updating feed: Unable to parse payload: 1: 11: error: unknown field: action.
```

</details>

update: Code related issues in the new implementation

**Comment by MiguelazoDS** [comment]
# Update 

From the comment here https://github.com/wazuh/wazuh/issues/36398#issuecomment-4579070895, one case was not taken into account 

``` console
  replace_vendor = 0,
  replace_product,
  replace_vendor_if_matches,
  replace_product_if_matches,
  set_version_if_matches,
  replace_sw_edition_if_product_matches,
  replace_msu_name_if_version_matches,
  ignore,
  check_hotfix,
  replace_msu_name,
  set_version_if_product_matches,
  set_targethw_if_product_matches,
  set_version_only_if_product_matches,
  set_targethw_only_if_product_matches,
  set_update_if_product_matches,
  set_update_only_if_product_matches
```

We do not need direct conditional to replace certain field, by defect it will replace if it is found. But for the Firefox case on Ubuntu (and probably many more) we have the same package from two different sources, and we only want to translate the one uses NVD. This seems a "lucky" case where either Firefox from PPA and Snap, both uses NVD, but that could not be always the case. 
In that scenarios may be worth to keep the 

```console
replace_vendor_if_matches
replace_product_if_matches
set_version_if_product_matches
```

**Comment by MiguelazoDS** [comment]
## Firefox case 

Originally, we avoided scan the Firefox package when the installation method is Mozilla PPA https://github.com/wazuh/wazuh/issues/28563 due to false positives 

Installing the package following this https://launchpad.net/%7Emozillateam/+archive/ubuntu/ppa?field.series_filter=jammy we get 

```console
{
  "architecture": "amd64",
  "description": "Safe and easy web browser from Mozilla",
  "format": "deb",
  "groups": "web",
  "install_time": " ",
  "location": " ",
  "multiarch": "",
  "name": "firefox",
  "priority": "optional",
  "size": 304442368,
  "source": " ",
  "vendor": "Ubuntu Mozilla Team <ubuntu-mozillateam@lists.ubuntu.com>",
  "version": "152.0.1+build1-0ubuntu0.24.04.1~mt1"
}
```

Installing from the repository, we get the same vendor but the version has different format

```console
{
  "architecture": "amd64",
  "description": "Transitional package - firefox -> firefox snap",
  "format": "deb",
  "groups": "oldlibs",
  "install_time": " ",
  "location": " ",
  "multiarch": "",
  "name": "firefox",
  "priority": "optional",
  "size": 123904,
  "source": " ",
  "vendor": "Ubuntu Mozilla Team <ubuntu-mozillateam@lists.ubuntu.com>",
  "version": "1:1snap1-0ubuntu5"
}
```

Following this guide here https://www.omgubuntu.co.uk/2022/04/how-to-install-firefox-deb-apt-ubuntu-22-04

```console
{
  "architecture": "amd64",
  "description": "Mozilla Firefox",
  "format": "deb",
  "groups": "web",
  "install_time": " ",
  "location": " ",
  "multiarch": "",
  "name": "firefox",
  "priority": "optional",
  "size": 321013760,
  "source": " ",
  "vendor": "Mozilla <release@mozilla.com>",
  "version": "152.0.1~build1"
}
```

From this I can see that the Firefox package is not actually installed, although package provider reports the "1:1snap1-0ubuntu5" package. I'll ignore this for now, if required we could go back to this analysis https://github.com/wazuh/wazuh/issues/23656#issuecomment-4741952447

```console
root@39c656a429a3:/wazuh/src# firefox --version

Command '/usr/bin/firefox' requires the firefox snap to be installed.
Please install it with:

snap install firefox
```

**Comment by MiguelazoDS** [comment]
# Update 

Multiple tests were lost during the rebases, I need to restore them and validate the code.

**Comment by MiguelazoDS** [comment]
# Update 

The situation is not that bad as I mentioned here https://github.com/wazuh/wazuh/issues/23656#issuecomment-4869109068, there are some tests that could be added to guarantee the well-functioning. So far, everything seems to work as expected.
